### PR TITLE
Fix indexing behavior when input is zero-sized array

### LINKED ISSUE #503: IndexError when advanced-indexing zero-dim array
CuPy version: latest master (83f8a07)

Code:
```
import numpy, cupy
shape = ()

for xp in (numpy, cupy):
    print("--", xp.__name__)

    a = xp.ones(shape, numpy.float32)
    print(a.shape)
    print((a < 0).shape)
    print(a[a < 0].shape)
```

Result:
```
-- numpy
()
()
(0,)
-- cupy
()
()
Traceback (most recent call last):
  File "test-zero.py", line 10, in <module>
    print(a[a < 0].shape)
  File "cupy/core/core.pyx", line 1386, in cupy.core.core.ndarray.__getitem__ (cupy/core/core.cpp:30591)
    raise IndexError('too many indices for array')
IndexError: too many indices for array

```

### PR DESCRIPTION
**Pull request by okuta**
Fix #503

### CODE REVIEW

**Comment by okuta** [comment]
jenkins, test this please.

**Comment by okuta** [comment]
jenkins, test this please.

File: cupy/core/core.pyx [hunk 1]
```diff
@@ -3196,6 +3206,8 @@ cpdef ndarray _take(ndarray a, indices, li=None, ri=None, ndarray out=None):
             raise TypeError('Output dtype mismatch')
         if out.shape != out_shape:
             raise ValueError('Output shape mismatch')
+    if a.size == 0 and out.size != 0:
```
Comment (toslunar) [inline cupy/core/core.pyx]: The condition is different from the error message.  It seems the following case is not covered.
```
>>> cupy.empty((0, 0)).take([1], axis=0)
array([], shape=(1, 0), dtype=float64)
```

File: cupy/core/core.pyx [hunk 2]
```diff
@@ -3121,20 +3128,23 @@ cpdef _prepare_mask_indexing_single(ndarray a, ndarray mask, Py_ssize_t axis):
     masked_shape = lshape + (n_true,) + rshape
 
     # When mask covers the entire array, broadcasting is not necessary.
-    if mask.ndim == a.ndim and axis == 0:
+    if mask._shape.size() == a._shape.size() and axis == 0:
```
Comment (toslunar) [inline cupy/core/core.pyx]: `.ndim` and `._shape.size()` are mixed in the function.  Could you explain your policy on the choices?

File: cupy/core/core.pyx (hunk 1 above)
Comment (okuta) [reply cupy/core/core.pyx]: Thanks. but, `cupy.take` behaves wrap mode for performance.

Comment (toslunar) [reply cupy/core/core.pyx]: In this example, wrap mode means `cupy.take` computes `a[1%0]`, doesn't it?

Comment (toslunar) [reply cupy/core/core.pyx]: I wrote the issue numpy/numpy#11677.

**Comment by toslunar** [comment]
Jenkins, test this please.

**Pull request merged**